# Phishing Website Detection using Machine Learning

## Notebook 02: Data Cleaning and Preprocessing

### Objective

The objective of this notebook is to clean the dataset and prepare it for machine learning. We will identify and handle data quality issues such as duplicate records, verify data types, and save a cleaned version of the dataset for use in subsequent notebooks.

---

## 1. Import Required Libraries

The following libraries are imported for data loading and preprocessing.

- **NumPy** – Numerical operations
- **Pandas** – Data manipulation

In [ ]:
import numpy as np
import pandas as pd

## 2. Load Dataset

Load the phishing dataset into a Pandas DataFrame to begin the preprocessing stage.

In [ ]:
df = pd.read_csv("../dataset/Zieni_Dataset for phishing detection.csv")

print("Dataset loaded successfully! ")

Dataset loaded successfully! 


## 3. Create a Working Copy

A copy of the dataset is created to ensure that the original data remains unchanged during preprocessing.

In [ ]:
df_clean = df.copy()

print("Copy created")

Copy created


## 4. Dataset Shape

Check the number of rows and columns before performing any preprocessing. This serves as a reference point for later comparison.

In [ ]:
print("Shape: ", df_clean.shape)


Shape:  (10000, 75)


## 5. Missing Value Analysis

Check whether any feature contains missing values that need to be handled before model training.

In [ ]:
df_clean.isnull().sum()

num_dots_url            0
num_hyph_url            0
num_underline_url       0
num_slash_url           0
num_questionmark_url    0
                       ..
num_percent_param       0
length_param            0
tld_in_param            0
num_param               0
phishing                0
Length: 75, dtype: int64

### Interpretation

- No missing values were found in the dataset.
- All features contain complete information.
- No imputation techniques are required.

---

## 6. Data Type Verification

Verify the data type of each feature to ensure the dataset is suitable for machine learning.

In [ ]:
df_clean.dtypes

num_dots_url            int64
num_hyph_url            int64
num_underline_url       int64
num_slash_url           int64
num_questionmark_url    int64
                        ...  
num_percent_param       int64
length_param            int64
tld_in_param            int64
num_param               int64
phishing                int64
Length: 75, dtype: object

## 7. Duplicate Record Analysis

Identify duplicate observations in the dataset before deciding whether they should be removed.

In [ ]:
duplicates = df_clean.duplicated().sum()

print(f"Duplicate Rows : {duplicates}")

Duplicate Rows : 5871


### Interpretation

- Duplicate records are present in the dataset.
- Duplicate samples may introduce bias during model training.
- The next step is to inspect these duplicates before removing them.


---

## 8. Investigating Duplicate Records

Before removing duplicate records, it is important to understand their impact on the dataset.

In this section, we will determine:

- The number of unique records.
- The dataset size after removing duplicates.
- Whether duplicate records belong to the same target class.

This analysis helps justify the decision to remove duplicate observations during preprocessing.

### **Step 1** : Check Dataset Size After Removing Duplicates

In [ ]:
df_clean.drop_duplicates().shape

(4129, 75)

### Interpretation

- After removing duplicate records, the dataset contains **4,129 unique observations**.
- Approximately **58.7%** of the original records were identified as exact duplicates.
- This indicates that a large portion of the dataset consists of repeated observations.

### **Step 2** : Check Class Distribution After Removing Duplicates

In [ ]:
df_clean.drop_duplicates()["phishing"].value_counts()

phishing
1    3063
0    1066
Name: count, dtype: int64

### Interpretation

- The class distribution changes significantly after removing duplicate records.
- The cleaned dataset contains:
  - **3,063 phishing websites**
  - **1,066 legitimate websites**
- Unlike the original dataset, the cleaned dataset is now **imbalanced**, with phishing websites representing the majority class.
- This suggests that duplicate records were not equally distributed across both classes.

---

## 9. Remove Duplicate Records

After investigating the duplicate observations, we now remove the exact duplicate rows from the dataset.

Removing duplicate records improves data quality and prevents the model from learning the same information multiple times.

In [ ]:
df_clean = df_clean.drop_duplicates()

df_clean.shape

(4129, 75)

### Interpretation

- Duplicate records have been successfully removed from the dataset.
- The cleaned dataset now contains **4,129 unique observations**.
- This dataset will be used for all further analysis and machine learning experiments.

---

## 10. Save the Cleaned Dataset

Save the cleaned dataset so that it can be reused in subsequent notebooks without repeating the preprocessing steps.

In [ ]:
df_clean.to_csv("../dataset/cleaned_phishing_dataset.csv", index=False)

print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.


# 11. Initial Findings

- The dataset was successfully loaded and inspected.
- No missing values were found.
- A total of **5,871 duplicate records** were identified.
- After removing duplicates, **4,129 unique observations** remained.
- The removal of duplicate records changed the class distribution considerably.
- The cleaned dataset is now ready for exploratory data analysis and feature engineering.

# Conclusion

In this notebook, the dataset was cleaned by identifying and removing duplicate records while preserving the original dataset.

The preprocessing revealed that duplicate observations had a significant impact on the dataset size and class distribution. After cleaning, the dataset consists of unique observations and is ready for further analysis.

The next notebook will focus on **Exploratory Data Analysis (EDA)** to understand feature distributions, relationships, and patterns within the cleaned dataset.